packages/go_router/lib/src/go_route_information_parser.dart

```diff
@@ -4,20 +4,441 @@
+    assert(() {
+      _debugLogKnownRoutes();
+      return true;
+    }());
+  }
+
+  /// List of top level routes used by the go router delegate.
+  final List<GoRoute> routes;
+
+  /// The limit for the number of consecutive redirects.
+  final int redirectLimit;
+
+  /// Top level page redirect.
+  final GoRouterRedirect topRedirect;
+
+  /// A debug property to assert [GoRouteInformationProvider] is in use along
+  /// with this parser.
+  ///
+  /// An assertion error will be thrown if this property set to true and the
+  /// [GoRouteInformationProvider] is not in use.
+  ///
+  /// Defaults to false.
+  final bool debugRequireGoRouteInformationProvider;
+
+  final Map<String, String> _nameToPath = <String, String>{};
+
+  void _cacheNameToPath(String parentFullPath, List<GoRoute> childRoutes) {
+    for (final GoRoute route in childRoutes) {
+      final String fullPath = concatenatePaths(parentFullPath, route.path);
+
+      if (route.name != null) {
+        final String name = route.name!.toLowerCase();
+        assert(!_nameToPath.containsKey(name),
+            'duplication fullpaths for name "$name":${_nameToPath[name]}, $fullPath');
+        _nameToPath[name] = fullPath;
+      }
+
+      if (route.routes.isNotEmpty) {
+        _cacheNameToPath(fullPath, route.routes);
+      }
+    }
+  }
+
+  /// Looks up the url location by a [GoRoute]'s name.
+  String namedLocation(
+    String name, {
+    Map<String, String> params = const <String, String>{},
+    Map<String, String> queryParams = const <String, String>{},
+  }) {
+    assert(() {
+      log.info('getting location for name: '
+          '"$name"'
+          '${params.isEmpty ? '' : ', params: $params'}'
+          '${queryParams.isEmpty ? '' : ', queryParams: $queryParams'}');
+      return true;
+    }());
+    assert(_nameToPath.containsKey(name), 'unknown route name: $name');
+    final String path = _nameToPath[name]!;
+    assert(() {
+      // Check that all required params are present.
+      final List<String> paramNames = <String>[];
+      patternToRegExp(path, paramNames);
+      for (final String paramName in paramNames) {
+        assert(params.containsKey(paramName),
+            'missing param "$paramName" for $path');
+      }
+
+      // Check that there are no extra params
+      for (final String key in params.keys) {
+        assert(paramNames.contains(key), 'unknown param "$key" for $path');
+      }
+      return true;
+    }());
+    final Map<String, String> encodedParams = <String, String>{
+      for (final MapEntry<String, String> param in params.entries)
+        param.key: Uri.encodeComponent(param.value)
+    };
+    final String location = patternToPath(path, encodedParams);
+    return Uri(path: location, queryParameters: queryParams).toString();
+  }
+
+  /// Concatenates two paths.
+  ///
+  /// e.g: pathA = /a, pathB = c/d,  concatenatePaths(pathA, pathB) = /a/c/d.
+  static String concatenatePaths(String parentPath, String childPath) {
```

Comment:
Should this be private? Do customers need to call this API?

Reply:
Since the `go_route_information_parser.dart` library isn't exported by the `go_router.dart` library, this function isn't visible when you `import package:go_router/go_router.dart` as a user. 

One reason to keep it public is for tests, if this function had tests it would need to be library-public so the test library could import and use it.